labgrid/driver/shelldriver.py

```diff
@@ -208,12 +206,27 @@ def _put_ssh_key(self, key):
                 if key['key'] == new_key['key']:
                     self.logger.info("Key already on target")
                     return
-            self.logger.info("Key not on target, mounting...")
-            self._run_check('echo "{}" > /tmp/keys'.format(keyline))
-            self._run_check('chmod 600 /tmp/keys')
-            self._run_check('mount --bind /tmp/keys ~/.ssh/authorized_keys')
-            self._run_check('chmod 700 ~/.ssh')
-            self._run_check('chmod 644 ~/.ssh/authorized_keys')
+
+        if test_write == 0 and read_keys == 0:
+            self.logger.debug("Key not on target and writeable, concatenating...")
+            self._run_check('echo "{}" >> ~/.ssh/authorized_keys'.format(keyline))
+            self._run_check("rm ~/.test")
+            return
+
+        if test_write == 0:
+            self.logger.debug("Key not on target, no ssh directory and writeable, concatenating...")
```

Comment:
read_keys==0 only means that authorized_keys does not exist, the .ssh dir may exist anyway. Also, why are you concatenating in this case?

Reply:
Concatenating is wrong here, copy paste error from above. I added a check for the directory.